src/poprox_recommender/components/diversifiers/locality_calibration.py

```diff
@@ -0,0 +1,76 @@
+from collections import defaultdict
+
+import torch as th
+
+from poprox_concepts import Article, ArticleSet, InterestProfile
+from poprox_recommender.diversifiers import Calibrator, compute_kl_divergence
+from poprox_recommender.lkpipeline import Component
+from poprox_recommender.topics import extract_locality, normalized_topic_count
+
+
+# Locality Calibration uses MMR
+# to rerank recommendations according to
+# locality calibration
+class LocalityCalibrator(Component, Calibrator):
+    def __init__(self, theta: float = 0.1, num_slots=10):
+        super.__init__(theta, num_slots)
+
+    def __call__(self, candidate_articles: ArticleSet, interest_profile: InterestProfile) -> ArticleSet:
+        normalized_locality_prefs = normalized_topic_count(interest_profile.click_locality_counts)
+
+        if candidate_articles.scores is not None:
+            article_scores = th.sigmoid(th.tensor(candidate_articles.scores))
+        else:
+            article_scores = th.zeros(len(candidate_articles.articles))
+
+        article_scores = article_scores.cpu().detach().numpy()
+
+        article_indices = self.calibration(
+            article_scores,
+            candidate_articles.articles,
+            normalized_locality_prefs,
+            self.theta,
+            topk=self.num_slots,
+        )
+        return ArticleSet(articles=[candidate_articles.articles[int(idx)] for idx in article_indices])
+
+    def calibration(self, relevance_scores, articles, preferences, theta, topk) -> list[Article]:
```

Comment:
Request: This function could also return a list of flags for articles that have been swapped from the original relevance_scores list for easy manipulation of those articles' slugs in a later stage.

Reply:
Relevance scores is sorted in a sense, but not in order descending by score. The order matches the list of candidate articles, so that the same position in the list of articles and in the list of scores correspond to each other.

In theory, this calibration component could compute the top-k list and do some kind of comparison to generate the kind of flags you're talking about, but it currently generates its own list of recs directly from the scores rather than operating on the top-k list and swapping things around.

If you want to identify which positions of the list are different from the plain top-k list, I'd suggest using this component, [the TopkRanker component](https://github.com/CCRI-POPROX/poprox-recommender/blob/main/src/poprox_recommender/components/rankers/topk.py), and comparing the output of the two in a new `GenerateExplanations` component.

Reply:
Thanks for pointing us to the `TopRanker` component; Indeed there should be a new `GenerateExplanations` component where we'll apply the LLM logic. We'll discuss that with the team.